Avoid measurement constraints

Robbybp opened the pull request:
## Summary/Motivation:
I made some changes to avoid needing measurement constraints (and measurement error variables), and also some experimental changes to streamline the construction of a tracking objective from a time-varying target trajectory.

Take a look when you get a chance.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

Robbybp commented:
@kuanhanl I just pushed some changes to avoid creating "model disturbance" variables as well. Instead, I just penalize the residual of the differential equations (at the sample points).

Robbybp commented on pyomo/contrib/mpc/examples/cstr/run_mhe.py:
```diff
@@ -129,68 +115,20 @@ def run_cstr_mhe(
+        variables=measured_variables,
     )
-
-    #
-    # Initialize measurements to initial values of measured variables
-    #
-    for index, var in enumerate(measured_variables):
-        for spt in m_estimator.sample_points:
-            esti_blo.measurement_variables[index, spt].set_value(var[spt].value)
```
This initialization now happens when the measurement parameters are constructed.

kuanhanl replied:
Okay!

kuanhanl commented on pyomo/contrib/mpc/modeling/cost_expressions.py:
```diff
@@ -246,3 +277,52 @@ def tracking_rule(m, t):
+):
+    if weight_data is None:
+        weight_data = ScalarData(
+            ComponentMap((var, 1.0) for con in constraints)
```
It should be `ComponentMap((con, 1.0) for con in constraints)`.

kuanhanl commented on pyomo/contrib/mpc/modeling/cost_expressions.py:
```diff
@@ -246,3 +277,52 @@ def tracking_rule(m, t):
+                    " of constraint %s" % con[t].name
+                )
+        resid_expr_list.append(resid_expr_dict)
+    # NOTE: In KH's implementation, using error vars enforces that constraint
+    # residuals are constant throughout a sampling period. Is this necessary?
+    # Supposing that it is, we can achieve the same thing by imposing piecewise
+    # constant constraints on these expressions.
```
I think it is not necessary but more like a traditional thing to assume the disturbance (or the term "residual" you use here) is constant throughout a sampling period. For example, in the model dynamics `x_{k+1} = f(x_k, u_k, w_k)` we say `w_k` may be different between different time steps but constant at each time step (t_k).

kuanhanl commented on pyomo/contrib/mpc/modeling/cost_expressions.py:
```diff
@@ -246,3 +277,52 @@ def tracking_rule(m, t):
+    for con in constraints:
+        if not weight_data.contains_key(con):
+            raise KeyError(
+                "Tracking weight does not contain a key for"
+                " constraint %s" % con
```
Not the tracking weight here. Maybe residual/disturbance weight?

kuanhanl commented on pyomo/contrib/mpc/modeling/cost_expressions.py:
```diff
@@ -246,3 +277,52 @@ def tracking_rule(m, t):
+            expr = con[t].expr
+            if isinstance(expr, EqualityExpression):
+                resid_expr_dict[t] = (con[t].body - con[t].upper)
+            elif con.upper is None:
+                resid_expr_dict[t] = (con[t].lower - con[t].body)
+            elif con.lower is None:
+                resid_expr_dict[t] = (con[t].body - con[t].upper)
```
Do we need these cases? I thought we should raise the error if the model constraints are inequalities.

kuanhanl commented on pyomo/contrib/mpc/examples/cstr/run_mhe.py:
```diff
@@ -78,45 +74,35 @@ def run_cstr_mhe(
     m_estimator.estimation_block = pyo.Block()
     esti_blo = m_estimator.estimation_block
```
I prefer to put anything related to MHE under this block, but we can remove this block if it is not used.

kuanhanl commented on pyomo/contrib/mpc/examples/cstr/run_mhe.py:
```diff
@@ -78,45 +74,35 @@ def run_cstr_mhe(
-        esti_blo.disturbed_constraints,
+    m_estimator.disturbance_set = resid_set
+    m_estimator.residual_expr = resid
+    _, pwc_con = estimator_interface.get_piecewise_constant_constraints(
+        estimator_interface.slice_components(m_estimator.residual_expr),
+        sample_points,
```
This is great!

kuanhanl commented on pyomo/contrib/mpc/examples/cstr/tests/test_mhe.py:
```diff
@@ -83,10 +83,10 @@ def test_mhe_simulation(self):
 
         self.assertStructuredAlmostEqual(
-            pred_sim_data, sim_AB_data.get_data(), delta=1e-5
+            pred_sim_data, sim_AB_data.get_data(), reltol=1e-2
```
Shall we change `delta` to `reltol` in `test_mpc.py` as well??

kuanhanl commented on pyomo/contrib/mpc/interfaces/model_interface.py:
```diff
@@ -397,6 +416,53 @@ def get_tracking_cost_from_constant_setpoint(
+            # NOTE: Nondeterministic order in Python < 3.7
+            variables = [
+                self.model.find_component(key)
+                for key in setpoint_data.get_data().keys()
```
It should be `target_data` instead of `setpoint_data`.

kuanhanl commented on pyomo/contrib/mpc/interfaces/model_interface.py:
```diff
@@ -397,6 +416,53 @@ def get_tracking_cost_from_constant_setpoint(
+                    "Variables must be provided if we are using a Param or"
+                    " Expression as an argument"
+                )
+            target_data = TimeSeriesData.from_pyomo_components(
+                variables,
+                target_data,
+                time,
+                time_set=self.time,
+                context=context,
+            )
+        elif not isinstance(target_data, TimeSeriesData):
+            target_data = TimeSeriesData(*target_data, time_set=self.time)
```
Sometimes I'm really amazed how you can think of something super creative......
It's nice here to use `TimeSeriesData` to save measurement components.

Robbybp replied:
I might change how this constructor works to accept a list of time-only indexed parameters (or variables, expressions, etc.) as the second argument instead of a single parameter indexed by two sets. This would be slightly nicer for components that don't come from one of our "modeling" functions, but would be slightly more onerous for components that do as we would have to call `interface.slice_components` on them...

kuanhanl commented on pyomo/contrib/mpc/examples/cstr/run_mhe.py:
```diff
@@ -129,68 +115,20 @@ def run_cstr_mhe(
+    m_estimator.measurement_error_cost = cost
+
+    m_estimator.squred_error_disturbance_objective = pyo.Objective(expr=(
+        sum(m_estimator.measurement_error_cost.values())
+        + sum(
+            m_estimator.residual_expr[i, t]
+            for i in m_estimator.disturbance_set
+            for t in sample_points
```
I recommend adding some notes to explain why you use `.values()` to sum over the measurement expressions but you use two `for` to sum over the residual expressions. (I think the reason is we sum over sample points only but the residual expressions are defined at all time points)

Robbybp replied:
I think I'll update `measurement_error_cost` to be indexed by integers as well as time, then update it to use the same syntax as `residual_expr` here.

kuanhanl reviewed (commented):
I am pretty good with this branch. The objective function for MHE might be slightly messy and not straightforward compared to the formulation in the literature, but I think that's not a big problem.

However, after these reformulations, I wonder why we cannot get a "perfect" fit in this noise-free case.

Robbybp commented on pyomo/contrib/mpc/data/series_data.py:
```diff
@@ -34,6 +35,18 @@ class TimeSeriesData(_DynamicDataBase):
     """
 
+    @classmethod
+    def from_pyomo_components(
+        cls, keys, values, time, time_set=None, context=None
+    ):
+        # We assume values is an indexed Variable, Param, or Expression
+        # indexed by integer indices into the list of keys (variables)
+        # and the time set.
+        data = ComponentMap([
+            (var, [values[i, t] for t in time]) for i, var in enumerate(keys)
+        ])
```
With the `slice_components` method, this might not need to exist, as we can always construct as:
```
TimeSeriesData(
    ComponentMap(zip(keys, slice_components(values))),
    time,
)
```

Robbybp replied:
Actually, the object would be instantiated:
```python
TimeSeriesData(
    ComponentMap(zip(keys, [ref(slc) for slc in slice_components(values)])),
    time,
)
```
which is probably just onerous enough to justify having a constructor that does this for us. Note that unlike the keys, the values actually need to be Pyomo Components (rather than slices) so that they can be evaluated as expressions.

Robbybp replied:
Maybe the name `from_pyomo_component` would be better, as this constructor is treating `values` as a single component.

Robbybp commented on pyomo/contrib/mpc/modeling/cost_expressions.py:
```diff
@@ -34,6 +38,33 @@
 )
 
 
+def get_parameters_from_variables(
+    variables,
+    time,
+    ctype=Param,
+):
+    n_var = len(variables)
+    init_dict = {
+        (i, t): var[t].value for i, var in enumerate(variables) for t in time
+    }
+    var_set, comp = _get_indexed_parameters(
+        n_var, time, ctype=ctype, initialize=init_dict
+    )
+    return var_set, comp
```
This should probably live somewhere other than cost_expressions.py. Maybe just a param.py file? components.py?

Robbybp replied:
This should also have a method in the Interface class that (a) automatically uses the pre-specified time set and (b) allows CUIDs (etc.) as arguments by resolving them to variables via the model.

Robbybp commented on pyomo/contrib/mpc/modeling/cost_expressions.py:
```diff
@@ -246,3 +277,52 @@ def tracking_rule(m, t):
     return tracking_cost
+
+
+def get_constraint_residual_expression(
+    constraints,
+    time,
+    weight_data=None,
+    # TODO: Option for norm (including no norm)
+):
```
Same as above; this should have a method on the interface class.

Robbybp reviewed (commented):
Leaving myself some comments so I remember what I need to do on this PR. In addition, I think I prefer the API for constructing components that we use here - where we return a single component indexed by an integer set as well as time. I think I will update the component construction methods in the upstream branch `contrib-mpc` to follow this pattern.

Robbybp commented:
In addition, since we are now using piecewise constant constraints for more than just inputs, I should probably rename the file containing these constraints to just "constraints.py"

Robbybp commented:
I've made some updates to the "main PR branch" `contrib-mpc` that I think will allow things to be more consistent for MHE. I'll merge these changes into our branches here then work on the updates I mention above.

emma58 committed:
Regestering the outer repn transformation under contrib.piecewise also

emma58 committed:
Moving ConfigDict to base class

jsiirola committed:
Merge pull request #2708 from emma58/pw-linear-function

Adding contrib.piecewise package

jsiirola committed:
Merge branch 'main' into pw-linear-or

jsiirola committed:
Simplify / standardize polynomial degree

jsiirola committed:
Simplify / standardize _to_string()

jsiirola committed:
Remove redundant code / simplify create_node_with_local_data

jsiirola committed:
Remove unnecessary derived class overrides

jsiirola committed:
NFC: comments / remove commented code

emma58 committed:
Adding reduced inner rep'n transformation and convex combination transformation

emma58 committed:
Merge branch 'main' into pw-linear-or

emma58 committed:
Merge branch 'pw-linear-or' of github.com:emma58/pyomo into pw-linear-or

emma58 committed:
Adding note about the simplification to get Vielma et al.'s multiple choice model from our multiple choice transformation

emma58 committed:
Fixing a comment

emma58 committed:
Just passing the log function directly in the test model

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
move CyIpoptNLP and CyIpoptProblemInterface to their own file and introduce a deferred import back into cyipopt_solver

Robbybp committed:
move cyipopt_interface.py to pynumero/interfaces/

emma58 committed:
Merge branch 'pw-linear-or' into pw-linear-ir-reduced

emma58 committed:
Fixing a lot of typos in reduced inner rep'n

Robbybp committed:
access cyipopt_available from cyipopt_interface module

emma58 committed:
We can use the PWLinearFunction's _points list to keep track of the extreme points and index the lambdas

emma58 committed:
Adding half-written tests (correct for log, but not for paraboloid yet)

Robbybp committed:
update imports to import CyIpoptNLP out of cyipopt_interfaces instead of cyipopt_solver

Robbybp committed:
call gc.collect before removing log files to try to make sure ipopt log file is closed

jsiirola committed:
Support comparing sequences of expressions

jsiirola committed:
Improve exception message

jsiirola committed:
Enforce list args in SumExpression __init__

jsiirola committed:
Simplify SumExpression create_node_with_local_data

jsiirola committed:
Fix ExternalFunxtionExpression PRECEDENCE

jsiirola committed:
Deprecate SumExpression.add()

jsiirola committed:
Remove unnecessary method overload

jsiirola committed:
NFC: improve comments

jsiirola committed:
NFC: fix typo

jsiirola committed:
Additional dispatcher tests

jsiirola committed:
Fix printing of non-numeric args for expression nodes without infix representation

jsiirola committed:
Test _balanced_parens

jsiirola committed:
Add additional testing of numeric_expr.py

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
Update test to track change in create_node_with_local_data

jsiirola committed:
Remove unreachable code, add missing call to value()

emma58 committed:
Finishing the paraboloid tests for the reduced inner repn transformation

jsiirola committed:
Remove unnecessary override

jsiirola committed:
Fix calculation of polynomial_degree(f(.)**0)

jsiirola committed:
Add optional optimizations for 0 in expression trees

...and standardize the order of argument processing, plus remove the
automatic optimization of f(.)**0 -> 1.

jsiirola committed:
Add tests for the 'zero' expression optimizations

jsiirola committed:
Check that all native_numeric_types are hashable

jsiirola committed:
NFC: improve docs

jsiirola committed:
NFC: fix typo, remove commented code

jsiirola committed:
Improve performance of LinearExpression.__init__

jsiirola committed:
Switch AMPLRepn.linear from list to dict

jsiirola committed:
Performance: faster printing of defined variables

jsiirola committed:
fix iteration over dict generators

jsiirola committed:
Fix typo

jsiirola committed:
Improve diff reporting (report actual baseline filename)

jsiirola committed:
Add the option to build againse pyomo-model-libraries branch

jsiirola committed:
Fix checkout of pyomo-model-libraries branch

jsiirola committed:
Reduce the size of the test problem slightly

jsiirola committed:
Merge pull request #2766 from emma58/pw-linear-or

contrib.piecewise: Adding transformation to outer representation GDP

jsiirola committed:
Fix for checking out pyomo-model-libraries

jsiirola committed:
Merge branch 'main' into nl-linear-dict

jsiirola committed:
Merge branch 'main' into pw-linear-ir-reduced

jsiirola committed:
Another attempt at identifying branch name

jsiirola committed:
Merge remote-tracking branch 'refs/remotes/me/nl-linear-dict' into nl-linear-dict

jsiirola committed:
Get the head (not base) ref

jsiirola committed:
Fall back on github.ref when not a PR

emma58 committed:
Merge pull request #2768 from emma58/pw-linear-ir-reduced

contrib.piecewise: Adding transformation to "reduced" inner representation GDP

jsiirola committed:
Merge pull request #2769 from jsiirola/nl-linear-dict

NL writer: improve linear datastructure

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
Merge branch 'main' into cyipopt-subclass

Robbybp committed:
call close before removing log file to attempt to remove lock

Robbybp committed:
add call to cyipopt.Problem.close at end of PyomoCyIpoptSolver.solve

Robbybp committed:
add a cyipopt.Problem.close call to a test that was missing it

mrmundt committed:
Reapply black to previously completed directories

mrmundt committed:
Update dReadtheDocs config to address deprecation

mrmundt committed:
Merge pull request #2780 from mrmundt/readthedocs-deprs

[Docs] Update ReadtheDocs Configuration

mrmundt committed:
Merge branch 'main' into black/reapply

mrmundt committed:
Merge pull request #2775 from mrmundt/black/reapply

Reapply `black` to previously completed directories

jsiirola committed:
Remove the use of weakrefs in SymbolMap

michaelbynum committed:
unify appsi solver logging

michaelbynum committed:
Merge branch 'main' into appsi_logging

mrmundt committed:
Apply `black` to `pyomo/common` directory `py` files (#2776)

* Apply black to pyomo/common directory py files

* Accidentally removed a necessary space

* Restoring pyomo/core as they are not supposed to be in this change

* Fix missed string merges

* Fix odd indentation issue

mrmundt committed:
Apply `black` to `kernel` and `gdp` directory `py` files (#2784)

* Apply black to kernel and gdp directory py files

* Remove unnecessary comment

mrmundt committed:
Apply `black` to `pyomo/dataportal` directory `py` files (#2778)

* Apply black to pyomo/dataportal directory py files

* Apply odd mixed string fix

Robbybp committed:
fix error message

mrmundt committed:
Apply black to duality, mpec, and environ directory py files (#2779)

Robbybp committed:
test case with indexed block

mrmundt committed:
Apply black to pyomo/opt directory py files (#2782)

Robbybp committed:
make sure pyomo_nlp is available before checking isinstance(, PyomoNLP)

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into incidence-indexed-block

michaelbynum committed:
Merge pull request #2787 from michaelbynum/appsi_logging

Appsi logging

jsiirola committed:
Merge branch 'main' into symbolmap-no-weakref

jsiirola committed:
Merge branch 'main' into symbolmap-no-weakref

jsiirola committed:
NLv2: Bugfix for shared named expressions

jsiirola committed:
Merge pull request #2789 from Robbybp/incidence-indexed-block

Test indexed block in Incidence Analysis

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
Add error checking to addSymbol

jsiirola committed:
Fix inconsistencies in symbol map interface

jsiirola committed:
Removing an unnecessary test (that will always be true)

jsiirola committed:
Adding additional SymbolMap testing

jsiirola committed:
Reverting 4071f2f

jsiirola committed:
NFC: Updating comments

jsiirola committed:
Merge pull request #2790 from jsiirola/nl-shared-expr-fix

NLv2: Bugfix for shared named expressions

mrmundt committed:
Set up git-blame-ignore file

jsiirola committed:
Merge pull request #2791 from jsiirola/symbolmap-no-weakref

SymbolMap: remove the use of weakrefs

mrmundt committed:
Add black linting action

mrmundt committed:
Adjust contrib docs for black/spell check expectations

mrmundt committed:
Add false typos and ignore a file type

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
call close() on user-created problem object rather than private _problem attribute

Robbybp committed:
note that close must be called to force Ipopt to release its lock on the log file

Commit:
Fix community.rst

Import repeated

mrmundt committed:
Merge pull request #2792 from DavidDB33/patch-1

Fix community.rst

mrmundt committed:
Replace pandas append with concat

mrmundt committed:
Change concat usage for pandas 2.0.0

mrmundt committed:
Merge pull request #2795 from mrmundt/pandas-2.0.0

Update to Parmest for pandas 2.0.0 release

mrmundt committed:
Fix typos in all dirs to which black has been applied already

mrmundt committed:
Apply `black` to `pyomo/contrib` directory `py` files (#2781)

* Apply black to pyomo/contrib directory py files

* Apply black to residual pyomo/contrib files

* Clean up mixed up branch

* Apply fixes to odd comment corrections

* Address one missed awkward indentation

mrmundt committed:
Add SHA for black/contrib

mrmundt committed:
Merge branch 'main' into auto-linting

mrmundt committed:
Merging main

mrmundt committed:
Apply typo fixes to contrib

mrmundt committed:
Reapply black to some dual modified files

mrmundt committed:
Temporarily turn off linting to verify changes

mrmundt committed:
Manually correct missed typos

emma58 committed:
Filtering out degenerate simplices after Delaunay triangulation, raising a DeveloperError when it appears that failed, warning user about points dropped in the triangulation

emma58 committed:
Blackification

mrmundt committed:
Apply `black` to `dae` and `network` directory `py` files (#2783)

* Apply black to dae and network directory py files

* NFC: cleaning up comment formatting

---------

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

mrmundt committed:
Merge branch 'main' into auto-linting

mrmundt committed:
Apply typo correctionss post DAE/Network black merge

mrmundt committed:
Apply `black` to `pyomo/core` directory `py` files (#2777)

* Apply black to pyomo/core directory py files

* NFC: cleaning up some comment/string formatting

---------

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

mrmundt committed:
Merge remote-tracking branch 'origin' into auto-linting

mrmundt committed:
Apply typo fixes and last pass at black formatting

mrmundt committed:
Turn black linting on for PRs

mrmundt committed:
Change black arguments

mrmundt committed:
Apply black to residual examples aftering converting to Python3

mrmundt committed:
Look at file directory

mrmundt committed:
Lock black to v22.6

emma58 committed:
Adding tests for degenerate simplices and redundant points

emma58 committed:
Fixing a typo

emma58 committed:
Adding argument for tabular data: a dictionary mapping points to nonlinear function values

emma58 committed:
NFC: applying black

emma58 committed:
Removing a redundant line

emma58 committed:
NFC: updating comments

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

conflicts were resolved by taking the cyipopt-subclass files and running
black on them.

Robbybp committed:
run black on files I missed

Robbybp committed:
re-run black, now with --skip-string-normalization

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

emma58 committed:
Sorting the simplices since apparently Delaunay behaves differently on Windows

emma58 committed:
Sorting the simplex indices when we store them in the pw-linear function, which means that later sorting the simplices means something.

jsiirola committed:
Update magic constant for visitor tests for python 3.11.2

emma58 committed:
Changing tests to just verify that there are 6 full-dimensional simplices inside of each of the 8 cubes of the grid.

emma58 committed:
Merge branch 'main' into pw-linear-fix

mrmundt committed:
Merge pull request #2799 from jsiirola/fix-visitor-tests-3.11

Update magic constant for visitor tests for python 3.11.2

mrmundt committed:
Apply black 23.3.0 - adding white space

mrmundt committed:
Unpin black version

mrmundt committed:
Merge branch 'main' into auto-linting

mrmundt committed:
Add mass black reapplication to ignore list

mrmundt committed:
Update instructions for local configuration, not global

mrmundt committed:
Fix CHANGELOG line to reflect typo change

jsiirola committed:
Merge branch 'main' into pw-linear-fix

mrmundt committed:
Update contributer's guide with full black command

mrmundt committed:
Residual awkward line combinations from black

mrmundt committed:
Apply black change

mrmundt committed:
Apply comments per @jsiirola

mrmundt committed:
Standardization of word usage

mrmundt committed:
Merge pull request #2800 from mrmundt/auto-linting

Auto-Linting: Spelling Black Style Checker

emma58 committed:
It's not always my fault if the simplices are degenerate--keeping track of where to place the blame and throwing either a ValueError or DeveloperError accordingly.

emma58 committed:
Merge branch 'pw-linear-fix' of github.com:emma58/pyomo into pw-linear-fix

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
Addressing reviewer comments

jsiirola committed:
Applying black

jsiirola committed:
NFC: fixing spelling

emma58 committed:
Skipping test when numpy not available

emma58 committed:
Merge branch 'main' into pw-linear-fix

emma58 committed:
Merge branch 'main' into pw-linear-fix

emma58 committed:
Merge branch 'pw-linear-fix' of github.com:emma58/pyomo into pw-linear-fix

jsiirola committed:
Correct implementation of polynomial_degree() for **0

emma58 committed:
Changing logging to be in __name__

emma58 committed:
merging main

emma58 committed:
No typos here

jsiirola committed:
Updating tests for **0

jsiirola committed:
Merge pull request #2797 from emma58/pw-linear-fix

contrib.piecewise: Fixing a bug caused by degenerate simplices

emma58 committed:
Resolving merge conflict

emma58 committed:
Passing dimension through so that the we know it to filter the simplices from the Delaunay triangulation

jsiirola committed:
Improve formatting for DeveloperError, MouseTrap messages

jsiirola committed:
Switch signature for sympy _operatorMap dispatcher

jsiirola committed:
Switch Sympy2PyomoVisitor to use public sympy API

jsiirola committed:
Update Pyomo2SympyVisitor to resolve NPV expressions to constants

jsiirola committed:
Raise an exception when converting expressions with ExternalFunctions to sympy

jsiirola committed:
Switch cnf_walker to use the visitors from sympy_tools

jsiirola committed:
NFC: comment updates

jsiirola committed:
logical_to_linear: raise exception for logical constraints with relational expression nodes

jsiirola committed:
Adding tests

jsiirola committed:
Refining MouseTrap message

jsiirola committed:
Improve warning when dynamically registering numeric types

jsiirola committed:
Resolving test failures

jsiirola committed:
NFC: fix typos

jsiirola committed:
Merge pull request #2805 from jsiirola/improved-error-formatting

Improve formatting for DeveloperError, MouseTrap messages

jsiirola committed:
Merge branch 'main' into sympy-walker-updates

jsiirola committed:
Fix signature for external function callback

jsiirola committed:
Add logical_to_linear tests

jsiirola committed:
NFC: fix typo

jsiirola committed:
Merge pull request #2798 from emma58/pw-linear-funct-from-tabular-data-really

contrib.piecewise: Construct PiecewiseLinearFunction from tabular data

jsiirola committed:
Updating imports to reflect TPL from cython builds

jsiirola committed:
Merge pull request #2806 from jsiirola/sympy-walker-updates

Improve conversions between Pyomo and Sympy expressions

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
update formatting for consistency with black

Robbybp committed:
fix indentation

Robbybp committed:
comma at end of arg list

Robbybp committed:
remove whitespace

jsiirola committed:
NLv2: resolve NPV constants in LinearExpression to native type

jsiirola committed:
Improve initialization of LinearExpression objects

jsiirola committed:
Remove portability code for python < 3.7

jsiirola committed:
Additional LinearExpression NLv2 test

jsiirola committed:
NPC: apply black

jsiirola committed:
Fix index math error

jsiirola committed:
Revert change in exception type from #2806

jsiirola committed:
Support fallback to numeric differentiation

jsiirola committed:
NFC: apply black

jsiirola committed:
Merge pull request #2811 from jsiirola/fix-nlv2-npv-linear-const

NLv2: resolve NPV constants in LinearExpressions

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
add flag and check to catch if solve is calledwhen cyipopt.Problem.__init__ has not been called

blnicho committed:
NFC: Fixing string formatting

blnicho committed:
Merge pull request #2760 from Robbybp/cyipopt-subclass

Refactor CyIpopt interface to subclass `cyipopt.Problem`

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
NFC: Apply black

blnicho committed:
Merge pull request #2722 from jsiirola/expr-multiple-dispatch

Rework expression generation to leverage multiple dispatch

jsiirola committed:
GHA: link test is a prereq only for PR/main builds

jsiirola committed:
Fix GHA syntax

michaelbynum committed:
Merge pull request #2818 from jsiirola/gha-lint-branch

GHA: lint test is a prereq only for PR/main builds

jsiirola committed:
Merge branch 'main' into calc-var-value-diff-mode

jsiirola committed:
Adding tests for external functions

jsiirola committed:
Split external function test into 2 tests

jsiirola committed:
missing import

jsiirola committed:
Correct import location

michaelbynum committed:
fix cyipopt interface when load_solutions=False

jsiirola committed:
Merge pull request #2812 from jsiirola/calc-var-value-diff-mode

Improve robustness of `calculate_variable_from_constraint()`

michaelbynum committed:
merge main

michaelbynum committed:
run black

MLopez-Ibanez committed:
[DOC] Remove instructions for python <= 3.0

because Pyomo requires Python >= 3.7.

blnicho committed:
Adding blank line removed from doctest by mistake

mrmundt committed:
Merge pull request #2822 from MLopez-Ibanez/patch-1

[DOC] Remove instructions for python <= 3.0

emma58 committed:
Switching from checking determinant to checking matrix rank when filtering for degenerate simplices, because it uses a tolerance by default.

emma58 committed:
Adding a test where the given points will produce a simplex that is not full-dimensional, but the determinant is (slightly) non-zero, so we need a tolerance when we filter the simplices.

emma58 committed:
Removing unncecessary import

emma58 committed:
Removing some sneaky debugging

michaelbynum committed:
Merge pull request #2820 from michaelbynum/load_solutions

fix cyipopt interface when load_solutions=False

emma58 committed:
Merge pull request #2826 from emma58/pw-linear-fix

contrib.piecewise: Adding tolerance when filtering degenerate simplices

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into contrib-mpc-mhe-mutparam
